API/Intro.md

```diff
@@ -0,0 +1,151 @@
+
+```json
+// GET api/songs/123
+{
+  "id": 123,
+  "name": "Modern Love",
+  "artist_id": 491,
+  "album_id": 528
+}
+```
+
+The API developer decided that instead of serializing artist and album names, they want to return their IDs. Technically, this is not a bad decision, it's quite a common change actually. However, what they failed to do is communicate this change to their clients in advance, so that clients can update their code to support the new response.
+
+Clients have expectations about APIs. If you tell clients that an endpoint returns data in a particular format, they will start to build their code around it. An API changed without prior notice breaks those assumptions and client applications which depend on them.
+
+Fortunately, the remedy is simple: communicate with your clients. Anticipate that a change might break client applications and tell them about it in advance.
+
+### Consistent design
+
+Let's imagine an API endpoint which returns country data:
+
+```json
+// GET api/countries/hrv
+{
+  "Ime": "Croatia",
+  "nativeName": "Hrvatska",
+  "top_level_domain": ".hr"
+}
+```
+
+What's the issue here? As you can see, the API is using different naming conventions for attributes:
+
+* `Ime` - Croatian PascalCase
+* `nativeName` - English camelCase
+* `top_level_domain` - English snake_case
+
+The problem is not the choice of cases (every option is good), nor the language, but the inconsistency. When building an API, pick one style and stick with it.
+
+This doesn't apply just to request and response bodies, URLs are also subject to consistency rules, e.g. endpoints `/api/country_data` and `/api/currency-data` are bad design because they're mixing snake_case and kebab-case.
+
+### Following conventions and standards
+
+Here is an example of an endpoint which returns album data:
+
+```json
+// GET api/albums/123
+{
+  "name": "Elephant",
+  "explicit": "false",
+}
+```
+
+The API serializes the attribute `explicit` which tells the client if the album contains strong language. This has been implemented as a boolean field, however the API developer decided to stringify the boolean, so instead of returning `false`, they return `"false"`.
+
+### Performant
+
+As the project grows, the user base will also grow, thus increasing the amount of requests sent to your server. Without a fast and optimized API, the response times will sky-rocket and the user experience will plummet, which might cause users to delete their accounts and abandon your product.
+
+There are many ways to improve the performance of an API, from caching to database sharding, but please be aware that the amount of information fetched from the database and wrapped in an API response should be as small as it can be, so increasing the speed of SQL queries and avoiding any [performance issues](https://infinum.com/handbook/books/devproc/general-coding-practices/api-design#performance-issues) are crucial measures that should always be undertaken.
```

Comment:
```suggestion
There are many ways to improve the performance of an API, from caching to database sharding, but please be aware that the amount of information fetched from the database and wrapped in an API response should be as small as it can be. Increasing the speed of SQL queries and avoiding any [performance issues](https://infinum.com/handbook/books/devproc/general-coding-practices/api-design#performance-issues) are crucial measures that should always be undertaken.
```